Support XDG Base directory specification for Linux

WhiredPlanck opened the pull request:
This PR will fix Issue https://github.com/Aegisub/Aegisub/issues/226. Adapted from https://github.com/aria2/aria2/commit/8bc1d37.

The basic idea just is:
- If $HOME/.aegisub exists, then the program will still read/write config/cache/data/state from there;
- If $HOME/.aegisub doesn't exist, then the program will read/write config/cache/data/state from corresponding XDG base Directory.

The original PR is here: https://github.com/wangqr/Aegisub/pull/132

witchymary commented:
I'm pretty sure this will still run into the issues I fixed here: https://github.com/arch1t3cht/Aegisub/pull/160/commits/b5a071b15d2f0c8041341ccdb7427cbde3b187e9

Also worth noting that this will break regardless due to this line in Dependency Control (https://github.com/TypesettingTools/DependencyControl/blob/9b8d615f18e295f4777ea606a7173465bdaf140d/modules/DependencyControl/Record.moon#L44). It's need to be changed in there as well

zefr0x commented on libaegisub/unix/path.cpp:
```diff
@@ -70,12 +76,32 @@ namespace agi {
+	agi::fs::path old_root = home/".aegisub";
+
+	agi::fs::path xdg_config = xdg_dir("XDG_CONFIG_HOME", home/".config");
+	agi::fs::path xdg_cache = xdg_dir("XDG_CACHE_HOME", home/"cache");
```
```suggestion
	agi::fs::path xdg_cache = xdg_dir("XDG_CACHE_HOME", home/".cache");
```

WhiredPlanck replied:
Thanks!

WhiredPlanck commented:
> I'm pretty sure this will still run into the issues I fixed here: [arch1t3cht@b5a071b](https://github.com/arch1t3cht/Aegisub/commit/b5a071b15d2f0c8041341ccdb7427cbde3b187e9)
> 
> Also worth noting that this will break regardless due to this line in Dependency Control (https://github.com/TypesettingTools/DependencyControl/blob/9b8d615f18e295f4777ea606a7173465bdaf140d/modules/DependencyControl/Record.moon#L44). It needs to be changed in there as well

You're right, I will fix it.

WhiredPlanck committed:
Support XDG Base directory specification for Linux

Ref:
1. https://specifications.freedesktop.org/basedir-spec/latest
2. https://github.com/aria2/aria2/commit/8bc1d37

WhiredPlanck committed:
Fix compile issue caused by token to path mapping

Co-authored-by: witchymary <63314339+witchymary@users.noreply.github.com>

filip-hejsek commented on libaegisub/unix/path.cpp:
```diff
@@ -70,12 +76,32 @@ namespace agi {
+		SetToken("?state", old_root);
+	} else {
+		SetToken("?user", xdg_config/"aegisub");
+		SetToken("?local", xdg_cache/"aegisub");
```
I don't like pointing a directory named `?local` to a cache dir, the name `?local` suggests it's a local but persistent directory

filip-hejsek commented on libaegisub/unix/path.cpp:
```diff
@@ -41,6 +41,12 @@ std::string home_dir() {
 
+std::string xdg_dir(std::string_view token, agi::fs::path const& fallback) {
+	const char *env = getenv(token.data());
+	if (env) return env;
```
The environment variable must be ignored if it is empty or doesn't contain an absolute path
